### PR TITLE
[Improve] [CI]Add Sonar Check

### LINKED ISSUE #837: [Improve] Add Sonar Check
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

It can help us better check some possible problems with the project

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by liugddx**
close #837 
*(Please specify the module before the PR name: [core] ... or [flink] ...)*

### Purpose

Add sonar check for paimon.

1. Analyze paimon https://sonarcloud.io/projects/create cc @JingsongLi 

2. Create a GitHub Secret
   In paimon GitHub repository, go to [Settings > Secrets > Actions](https://github.com/apache/incubator-paimon/settings/secrets) and create a new secret with the following details:
2.1. In the Name field, enter `SONAR_TOKEN`.
2.2. In the Value field, enter token value.
2.3. In the Name field, enter `GITHUB_TOKEN`.(Needed to get PR information, if any) @JingsongLi 
2.4. In the Value field, enter token value.

https://docs.sonarcloud.io/advanced-setup/ci-based-analysis/github-actions-for-sonarcloud/

### Tests

*(List UT and IT cases to verify this change)*

### API and Format 

*(Does this change affect API or storage format)*

### Documentation

*(Does this change introduce a new feature)*

### CODE REVIEW

**Comment by liugddx** [comment]
@JingsongLi can we add sonar check for paimon?

**Comment by SteNicholas** [comment]
@liugddx, could you create an issue for sonar check? It's better to discuss whether to add sonar check in the issue.

**Comment by JingsongLi** [comment]
> @JingsongLi can we add sonar check for paimon?

I don't understand what it is for, can you explain this?

**Comment by liugddx** [comment]
> > @JingsongLi can we add sonar check for paimon?
> 
> I don't understand what it is for, can you explain this?

Sonar can help us find some bugs or performance problems. Such as https://sonarcloud.io/project/overview?id=apache_incubator-doris

**Comment by JingsongLi** [comment]
> > > @JingsongLi can we add sonar check for paimon?
> > 
> > 
> > I don't understand what it is for, can you explain this?
> 
> Sonar can help us find some bugs or performance problems. Such as https://sonarcloud.io/project/overview?id=apache_dubbo

Where to find bugs and performance problems? The sonar check will report errors?

**Comment by liugddx** [comment]
> > > > @JingsongLi can we add sonar check for paimon?
> > > 
> > > 
> > > I don't understand what it is for, can you explain this?
> > 
> > 
> > Sonar can help us find some bugs or performance problems. Such as https://sonarcloud.io/project/overview?id=apache_dubbo
> 
> Where to find bugs and performance problems? The sonar check will report errors?

Yes, sonar can set the quality gates and classify the detected problems.This is the overview page of hive.
image

**Comment by JingsongLi** [comment]
> > > > > @JingsongLi can we add sonar check for paimon?
> > > > 
> > > > 
> > > > I don't understand what it is for, can you explain this?
> > > 
> > > 
> > > Sonar can help us find some bugs or performance problems. Such as https://sonarcloud.io/project/overview?id=apache_dubbo
> > 
> > 
> > Where to find bugs and performance problems? The sonar check will report errors?
> 
> Yes, sonar can set the quality gates and classify the detected problems.This is the overview page of hive. image

The page is cool, but where is it?

**Comment by liugddx** [comment]
> > > > > > @JingsongLi can we add sonar check for paimon?
> > > > > 
> > > > > 
> > > > > I don't understand what it is for, can you explain this?
> > > > 
> > > > 
> > > > Sonar can help us find some bugs or performance problems. Such as https://sonarcloud.io/project/overview?id=apache_dubbo
> > > 
> > > 
> > > Where to find bugs and performance problems? The sonar check will report errors?
> > 
> > 
> > Yes, sonar can set the quality gates and classify the detected problems.This is the overview page of hive. image
> 
> The page is cool, but where is it?

it's here https://sonarcloud.io/project/overview?id=apache_hive

**Comment by JingsongLi** [comment]
@liugddx Hi, I think it is a good thing to do, what is missing here is a complete solution. How to integrate paimon to sonar. You should document this.

**Comment by JingsongLi** [comment]
You should write a complete solution in PR Purpose.

**Comment by liugddx** [comment]
> You should write a complete solution in PR Purpose.

I've added steps, some of which require admin intervention. Do you think this is ok?

**Comment by JingsongLi** [comment]
> > You should write a complete solution in PR Purpose.
> 
> I've added steps, some of which require admin intervention. Do you think this is ok?

This needs operation from INFRA, for example, https://issues.apache.org/jira/browse/INFRA-24358

**Comment by liugddx** [comment]
I see

**Comment by liugddx** [comment]
> > > You should write a complete solution in PR Purpose.
> > 
> > 
> > I've added steps, some of which require admin intervention. Do you think this is ok?
> 
> This needs operation from INFRA, for example, https://issues.apache.org/jira/browse/INFRA-24358

I created a new issue, is this okay?
https://issues.apache.org/jira/projects/INFRA/issues/INFRA-24434?filter=allissues

**Comment by JingsongLi** [comment]
> > > > You should write a complete solution in PR Purpose.
> > > 
> > > 
> > > I've added steps, some of which require admin intervention. Do you think this is ok?
> > 
> > 
> > This needs operation from INFRA, for example, https://issues.apache.org/jira/browse/INFRA-24358
> 
> I created a new issue, is this okay? https://issues.apache.org/jira/projects/INFRA/issues/INFRA-24434?filter=allissues

Let's waiting response from INFRA.

**Comment by liugddx** [comment]
> > > > > You should write a complete solution in PR Purpose.
> > > > 
> > > > 
> > > > I've added steps, some of which require admin intervention. Do you think this is ok?
> > > 
> > > 
> > > This needs operation from INFRA, for example, https://issues.apache.org/jira/browse/INFRA-24358
> > 
> > 
> > I created a new issue, is this okay? https://issues.apache.org/jira/projects/INFRA/issues/INFRA-24434?filter=allissues
> 
> Let's waiting response from INFRA.

Can you log in to sonarcloud.io? Chris needs to add you to the admin group. https://issues.apache.org/jira/browse/INFRA-24434?filter=-2

**Comment by liugddx** [comment]
Now paimon has joined the sonar check, please see
https://sonarcloud.io/project/overview?id=apache_incubator-paimon

**Comment by JingsongLi** [comment]
@liugddx Is all working as expected?

**Comment by liugddx** [comment]
> @liugddx Is all working as expected?

So far, it's working fine.
image

**Comment by JingsongLi** [comment]
@liugddx I can see there is sonarcloud bot now. And it works fine, so what is this `SonarCloud / build (pull_request)` for? What is difference between `SonarCloud / build (pull_request)` and `SonarCloud Code Analysis`?

**Comment by liugddx** [comment]
> @liugddx I can see there is sonarcloud bot now. And it works fine, so what is this `SonarCloud / build (pull_request)` for? What is difference between `SonarCloud / build (pull_request)` and `SonarCloud Code Analysis`?

My fault, what I saw was sonar's automatic check, https://docs.sonarcloud.io/advanced-setup/automatic-analysis/ and ci-based-analysis is not workl https://docs.sonarcloud.io/advanced-setup/ci-based-analysis/github-actions-for-sonarcloud/

**Comment by liugddx** [comment]
> > @liugddx I can see there is sonarcloud bot now. And it works fine, so what is this `SonarCloud / build (pull_request)` for? What is difference between `SonarCloud / build (pull_request)` and `SonarCloud Code Analysis`?
> 
> My fault, what I saw was sonar's automatic check, https://docs.sonarcloud.io/advanced-setup/automatic-analysis/ and ci-based-analysis is not workl https://docs.sonarcloud.io/advanced-setup/ci-based-analysis/github-actions-for-sonarcloud/

I don't know if I understand correctly,  We must merge pr into the trunk to see the effect. @JingsongLi 
https://issues.apache.org/jira/browse/INFRA-24434?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17709592#comment-17709592

**Comment by liugddx** [comment]
Also we have to turn off the automatic check. https://docs.sonarcloud.io/advanced-setup/automatic-analysis/#presence-of-a-properties-file. The close method is in the screenshot. 

image

**Comment by JingsongLi** [comment]
What are the drawbacks of automatic check? Why can't we use it?

**Comment by liugddx** [comment]
> What are the drawbacks of automatic check? Why can't we use it?

Currently, automatic analysis has the following limitations:

- Java can be analyzed automatically, but only a partial set of rules is supported.
- Code coverage information is not supported.

Of course, automatic check is also a very good choice. Let me know if automatic check is used, I will close the pr

**Comment by liugddx** [comment]
> > What are the drawbacks of automatic check? Why can't we use it?
> 
> Currently, automatic analysis has the following limitations:
> 
> * Java can be analyzed automatically, but only a partial set of rules is supported.
> * Code coverage information is not supported.
> 
> Of course, automatic check is also a very good choice. Let me know if automatic check is used, I will close the pr

- To be eligible for automatic analysis, your Java project must use either Maven or Gradle and the total amount of code in the project must be less than 10MB.
- Security vulnerability rules are not yet supported.
- Rules outside the Sonar way quality profile are not yet supported.
- Of the rules in the Sonar way quality profile, only a subset (about 80%) are currently supported.

**Comment by liugddx** [comment]
> Code coverage looks an attractive feature to me. Can code coverage be supported in this PR?

I will add coverage in this pr

File: .github/workflows/code-analysys.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: SonarCloud
+on:
+  schedule:
+    - cron: '0 1 * * *'
```
[the remark this answers is not in the document]
Comment (liugddx) [reply .github/workflows/code-analysys.yml:19]: > It is better to add `workflow_dispatch:`?

Done.

File: README.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # Paimon (Incubating)
 
+[![License](https://img.shields.io/badge/license-Apache%202-4EB1BA.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)
+[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=apache_incubator-paimon&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=apache_incubator-paimon)
+[![Stargazers over time](https://starchart.cc/apache/incubator-paimon.svg)](https://starchart.cc/apache/incubator-paimon)
```
Comment (JingsongLi) [inline README.md:5]: Can we remove `Stargazers over time`?

Comment (liugddx) [reply README.md]: > Can we remove `Stargazers over time`?

Done.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**